src/Samples/PsdSamples.cpp

```diff
@@ -358,7 +364,52 @@ int SampleReadPsd(void)
 			std::wstringstream layerName;
 			if (layer->utf16Name)
 			{
+				#ifdef _WIN32
+				//In Windows wchar_t is utf16
+				static_assert(sizeof(wchar_t) == sizeof(uint16_t));
 				layerName << reinterpret_cast<wchar_t*>(layer->utf16Name);
+				#else
+				//In Linux, wchar_t is utf32
+				//Convert code from https://stackoverflow.com/questions/23919515/how-to-convert-from-utf-16-to-utf-32-on-linux-with-std-library#comment95663809_23920015
+				//We cannot use C++11,so use macro instead of lambda
+				#define is_surrogate(uc) ( (uc - 0xd800u) < 2048u )
+				#define is_high_surrogate(uc) ( (uc & 0xfffffc00) == 0xd800 )
+				#define is_low_surrogate(uc)  ( (uc & 0xfffffc00) == 0xdc00 )
+				#define surrogate_to_utf32(high,low) ((high << 10) + low - 0x35fdc00)
+				static_assert(sizeof(wchar_t) == sizeof(uint32_t));
+
+				//Begin convert
+				size_t u16len = 0;
+				const uint16_t * cur = layer->utf16Name;
+				while(*cur != uint16_t('\0')){
+					cur++;
+					u16len++;
+				}
+				//Len it
+
+				const uint16_t * const end = layer->utf16Name + u16len;
+				const uint16_t * input = layer->utf16Name;
+				while (input < end) {
+					const uint16_t uc = *input++;
+					if (!is_surrogate(uc)) {
+						layerName << wchar_t(uc); 
+					} 
+					else {
+						if (is_high_surrogate(uc) && input < end && is_low_surrogate(*input)){
+							layerName << wchar_t(surrogate_to_utf32(uc, *input++));
+						}
+						else{
+							// Error
+							// Impossible
+							std::abort();
+						}
+					}
+				}
+				#undef is_surrogate
+				#undef is_high_surrogate
+				#undef is_low_surrogate
+				#undef surrogate_to_utf32
+				#endif

```

Comment:
I think it would be fine to use C++11 in this part of the sample code, which would probably simplify it a lot.
We could also make it so that when using a pre-C++11 compiler under Linux, the layername simply won't be output, so that the code doesn't **require** C++11, but rather uses it if available.

Reply:
I could not find the way to convert utf16 to utf32 with std library